sim/midas/targetutils/src/test/scala/RAMStyleHintSpec.scala

```diff
@@ -0,0 +1,99 @@
+    val writeEnable  = Input(Bool())
+  }
+
+  abstract class BaseMemoryModule(annotator: MemBase[_] => Unit) extends Module {
+    // Arbitrarily selected.
+    val dataBits = 64
+    val addrBits = 16
+    val io = IO(new MemoryModuleIO(addrBits, dataBits))
+
+    def mem: MemBase[_]
+    annotator(mem)
+  }
+
+
+  class SyncReadMemModule(annotator: MemBase[_] => Unit) extends BaseMemoryModule(annotator) {
+    // Lazy so this is elaborated  before annotator executes
+    lazy val mem = SyncReadMem(1 << addrBits, UInt(dataBits.W))
+    io.readData := mem.read(io.readAddress, io.readEnable)
+    when(io.writeEnable) { mem(io.writeAddress) := io.writeData }
+  }
+
+  // This uses a combinational-read chisel memory, but could be infered as a BRAM
+  // due to the pipelining of the data read out. So the check annotator still accepts these.
+  class CombReadMemModule(annotator: MemBase[_] => Unit) extends BaseMemoryModule(annotator) {
+    lazy val mem = Mem(1 << addrBits, UInt(dataBits.W))
+    when(io.readEnable)  { io.readData          := RegNext(mem(io.readAddress)) }
+    when(io.writeEnable) { mem(io.writeAddress) := io.writeData }
+  }
+
+  class WrapperModule(annotator: MemBase[_] => Unit) extends Module {
+    val a = IO(new MemoryModuleIO(64, 16))
+    val b = IO(new MemoryModuleIO(64, 16))
+
+    val modA = Module(new SyncReadMemModule(_ => Nil))
+    val modB = Module(new SyncReadMemModule(_ => Nil))
+    modA.io <> a
+    modB.io <> b
+
+    annotator(modA.mem)
+    annotator(modB.mem)
+  }
+
+  def annotateMem(style: RAMStyle)(mem: MemBase[_]): Unit = RAMStyleHint(mem, style)
+
+  def elaborateAndCollectXDCAnnos(mod: =>Module) =
+    elaborate(mod)._2.collect { case a: XDCAnnotation => a }
+
+  behavior of "RAMStyleHint"
+
+  // Sanity check that stuff passes through elaboration. 
+  it should "correctly annotate a chisel3.SyncReadMem as BRAM" in {
+    val annos = elaborateAndCollectXDCAnnos(new SyncReadMemModule(annotateMem(RAMStyles.BRAM)))
+    assert(annos.size == 1)
+  }
+
+  it should "correctly annotate a chisel3.SyncReadMem as URAM" in {
+    val annos = elaborateAndCollectXDCAnnos(new SyncReadMemModule(annotateMem(RAMStyles.ULTRA)))
+    assert(annos.size == 1)
+  }
+
+  it should "correctly annotate a chisel3.Mem as URAM" in {
+    val annos = elaborateAndCollectXDCAnnos(new CombReadMemModule(annotateMem(RAMStyles.ULTRA)))
+    assert(annos.size == 1)
+  }
+
+  it should "correctly annotate a chisel3.Mem as BRAM" in {
+    val annos = elaborateAndCollectXDCAnnos(new CombReadMemModule(annotateMem(RAMStyles.BRAM)))
+    assert(annos.size == 1)
+  }
+
+  it should "not trivially break deduplication" in {
+    val annos = elaborateAndLower(new WrapperModule(annotateMem(RAMStyles.ULTRA)))
+
+    val dedupResultAnnos = annos.collect { case a: firrtl.transforms.DedupedResult => a }
+    assert(dedupResultAnnos.size == 2)
+
+    val xdcAnnos = annos.collect { case a: XDCAnnotation => a }
+    assert(xdcAnnos.size == 2)
+  }
```

Comment:
Am I reading this correctly... you're annotating the duplicated `modA` and `modB` with the same type of RAMStyle and you don't expect dedup to collapse that into a single annotated module?  I'm confused why this is.

Should you test for the inverse case?  I would assume that would be where you anntate `modA` and `modB` with different `RAMStyle` and you would expect that to not be deduped, right?

Reply:
Dedup does deduplicate them (there would be no DedupResultAnnos otherwise; these show the remapping). The XDC annotations themselves point at instances of the deduped module instead, at least under current "annotation deduplication" semantics. 

AFAIK there is really no specification -- there is just the code -- that governs annotation behavior under deduplication, and as of 1.4 (i believe), the code changed to do this module -> instance style deduplication. 
In practice you want different behavior for different kinds of annotations. This was the source of much of our pain when doing the 1.4 bump. If you don't want the deduplication via mapping to instances, but instead want to combine the annotation into a single module-local annotation, i think you have to roll your own pass to do that.

In any case, the way these XDC annotations behave should work under both "styles" of deduplication since it can handle instance paths and module-local targets.